internal/controller/agentpool_controller_tokens.go

```diff
@@ -138,32 +85,37 @@ func (r *AgentPoolReconciler) createSecret(ctx context.Context, ap *agentPoolIns
 	}
 	if err := controllerutil.SetControllerReference(&ap.instance, s, r.Scheme); err != nil {
 		ap.log.Error(err, "Reconcile Agent Tokens", "msg", "failed to set controller reference")
-		return err
+		return nil, err
 	}
 	if err := r.Client.Get(ctx, getAgentPoolNamespacedName(&ap.instance), s); err != nil {
 		if errors.IsNotFound(err) {
 			ap.log.Info("Reconcile Agent Tokens", "msg", fmt.Sprintf("creating a new Kubernetes Secret %q", s.Name))
 			if err = r.Client.Create(ctx, s); err != nil {
 				ap.log.Error(err, "Reconcile Agent Tokens", "msg", fmt.Sprintf("failed to create a new Kubernetes Secret %q", s.Name))
-				return err
+				return nil, err
 			}
 			ap.log.Info("Reconcile Agent Tokens", "msg", fmt.Sprintf("successfully created a new Kubernetes Secret %q", s.Name))
-			return nil
+			return s, nil
 		}
 		ap.log.Error(err, "Reconcile Agent Tokens", "msg", fmt.Sprintf("failed to get Kubernetes Secret %q", s.Name))
-		return err
+		return nil, err
 	}
 
-	return nil
+	return s, nil
 }
 
 func (r *AgentPoolReconciler) reconcileAgentTokens(ctx context.Context, ap *agentPoolInstance) error {
 	ap.log.Info("Reconcile Agent Tokens", "msg", "new reconciliation event")
 
-	if err := r.createSecret(ctx, ap); err != nil {
+	s, err := r.createOrGetSecret(ctx, ap)
+	if err != nil {
 		ap.log.Error(err, "Reconcile Agent Tokens", "msg", fmt.Sprintf("failed to create a new Kubernetes Secret %s", agentPoolOutputObjectName(ap.instance.Name)))
 		return err
 	}
+	if s.Data == nil {
+		s.Data = make(map[string][]byte)
+	}
+	data := s.DeepCopy().Data
```

Comment:
Why do we need to DeepCopy the secret data?

Reply:
This keeps a snapshot of the original Secret data for comparison in the deferred function, helping us avoid an unnecessary update API call. A more elegant approach could involve temporarily using a Secret label as an indicator of an update. I will try it to see if it looks better.